tee-worker/litentry/core/assertion-build/src/a4.rs

```diff
@@ -0,0 +1,98 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+use crate::{Error, Result};
+use std::{
+	str::from_utf8,
+	string::{String, ToString},
+	vec,
+	vec::Vec,
+};
+
+use lc_stf_task_sender::MaxIdentityLength;
+use litentry_primitives::{
+	EvmNetwork, Identity, IdentityHandle, IdentityWebType, SubstrateNetwork, Web3Network,
+};
+use sp_runtime::BoundedVec;
+
+use lc_data_providers::graphql::{
+	GraphQLClient, VerifiedCredentialsIsHodlerIn, VerifiedCredentialsNetwork,
+};
+
+const LIT_TOKEN_ADDRESS: &str = "0xb59490aB09A0f526Cc7305822aC65f2Ab12f9723";
+
+pub fn build(
+	identities: BoundedVec<Identity, MaxIdentityLength>,
+	from_date: String,
+	mini_balance: f64,
+) -> Result<()> {
+	let mut client = GraphQLClient::new();
+
+	for identity in identities {
+		let mut network = VerifiedCredentialsNetwork::Polkadot;
+		if let IdentityWebType::Web3(web3_type) = identity.web_type {
+			match web3_type {
+				Web3Network::Substrate(SubstrateNetwork::Litentry) =>
+					network = VerifiedCredentialsNetwork::Litentry,
+				Web3Network::Substrate(SubstrateNetwork::Litmus) =>
+					network = VerifiedCredentialsNetwork::Litmus,
+				Web3Network::Evm(EvmNetwork::Ethereum) =>
+					network = VerifiedCredentialsNetwork::Ethereum,
+				_ => (),
+			}
+		};
+
+		if network == VerifiedCredentialsNetwork::Litentry
+			|| network == VerifiedCredentialsNetwork::Litmus
+			|| network == VerifiedCredentialsNetwork::Ethereum
+		{
+			let mut addresses: Vec<String> = vec![];
+			match identity.handle {
+				IdentityHandle::Address20(addr) =>
+					addresses.push(from_utf8(&addr).unwrap().to_string()),
+				IdentityHandle::Address32(addr) =>
+					addresses.push(from_utf8(&addr).unwrap().to_string()),
+				IdentityHandle::String(addr) =>
+					addresses.push(from_utf8(&addr).unwrap().to_string()),
+			}
+			let mut tmp_token_addr = String::from("");
+			if network == VerifiedCredentialsNetwork::Ethereum {
+				tmp_token_addr = LIT_TOKEN_ADDRESS.to_string();
+			}
+			let credentials = VerifiedCredentialsIsHodlerIn {
+				addresses,
+				from_date: from_date.clone(),
+				network,
+				token_address: tmp_token_addr,
+				mini_balance,
+			};
+			let is_hodler_out = client.verified_credentials_is_hodler(credentials);
+			if let Ok(_hodler_out) = is_hodler_out {
```

Comment:
If I understand it correctly, `addresses` can only have max one address, so if we have `n` linked identities for LIT, we will generate `n` VCs? It seems incorrect

Reply:
Not generate `n` VCs. Once one hodler_out matches, it will return. Line 93.

Reply:
I see, so the logic is if **any** of the linked identities is hodler, we'll have a VC saying `is_hodler` is true. And what if all of them are false? We still need to generate a VC.

It's fine that we leave it blank for now, but we desperately need a test soon :(

Reply:
`We still need to generate a VC.` then we need to add an `else` branch for this `if let ...`.

Reply:
I think it should be inside it, `if let` only checks if the output is meaningful but doesn't check if it's holder. Anyway it's another topic